### PR TITLE
Admin Add and Remove Perks

### PR DESCRIPTION
**Pull request by MSieghardt**
## About The Pull Request
<details>
<summary>
	Perk name is inputted manually, has to be the tailend of the path from https://github.com/sojourn-13/sojourn-station/blob/aca66e8cbbc9c32f48237728d838af8e87623db9/code/__DEFINES/perks.dm

So "sharp_mind" would be for the Sharpened Mind perk.

Might implement a list later, but this would require implementation with perks.dm to dynamically react when new perks are added. It's an admin command, doesn't need too much finesse at the moment.

Was tested in my own server, both direct command with Add-Perk and Remove-Perk work as well as using View Variables to add or remove perks. Not the most elegant picking as you have to type the perk name manually, but it'll work with any perk, both oddity and non-oddity ones. 

You're welcome, Kinni.
</summary>
<hr>

	
<hr>
</details>

## Changelog
:cl:

/:cl:

### CODE REVIEW

File: code/modules/admin/admin_verbs.dm [hunk 1]
```diff
@@ -555,6 +555,27 @@ ADMIN_VERB_ADD(/client/proc/man_up, R_ADMIN, FALSE)
 	log_admin("[key_name(usr)] told [key_name(T)] to man up and deal with it.")
 	message_admins("\blue [key_name_admin(usr)] told [key_name(T)] to man up and deal with it.", 1)
 
+ADMIN_VERB_ADD(/client/proc/perkadd, R_ADMIN, FALSE)
+/client/proc/perkadd(mob/T as mob in SSmobs.mob_list)
+	set category = "Fun"
+	set name = "Add Perk"
+	set desc = "Add a perk to a person."
+	var/datum/perk/perkname = input("What perk do you want to add?") in subtypesof(/datum/perk/)
+	T.stats.addPerk(perkname)
```
Comment (KathyRyals) [inline code/modules/admin/admin_verbs.dm:564]: As `input()` sleeps, you need to check that T wasn't QDELETED or marked for QDEL in the meantime. Use the defined : `QDELETED(var)` to check for it.

File: code/modules/admin/admin_verbs.dm [hunk 2]
```diff
@@ -555,6 +555,27 @@ ADMIN_VERB_ADD(/client/proc/man_up, R_ADMIN, FALSE)
 	log_admin("[key_name(usr)] told [key_name(T)] to man up and deal with it.")
 	message_admins("\blue [key_name_admin(usr)] told [key_name(T)] to man up and deal with it.", 1)
 
+ADMIN_VERB_ADD(/client/proc/perkadd, R_ADMIN, FALSE)
+/client/proc/perkadd(mob/T as mob in SSmobs.mob_list)
+	set category = "Fun"
+	set name = "Add Perk"
+	set desc = "Add a perk to a person."
+	var/datum/perk/perkname = input("What perk do you want to add?") in subtypesof(/datum/perk/)
```
Comment (KathyRyals) [inline code/modules/admin/admin_verbs.dm:563]: This needs to be cancelable, I think you could do it with 
`var/datum/perk/perkname = input("What perk do you want to add?") as null|anything in subtypesof(/datum/perk/)`
Needs to be tested. Don't forget to check if perkname is null afterward.

File: code/modules/admin/admin_verbs.dm [hunk 3]
```diff
@@ -555,6 +555,27 @@ ADMIN_VERB_ADD(/client/proc/man_up, R_ADMIN, FALSE)
...
+
+ADMIN_VERB_ADD(/client/proc/perkremove, R_ADMIN, FALSE)
+/client/proc/perkremove(mob/T as mob in SSmobs.mob_list)
+	set category = "Fun"
+	set name = "Remove Perk"
+	set desc = "Remove a perk from a person."
+	if (T.stats.perks.len ==0)
+		to_chat(usr, "Creature has no perks to remove")
+		return
+	var/datum/perk/perkname = input("What perk do you want to remove?") in T.stats.perks
```
Comment (KathyRyals) [inline code/modules/admin/admin_verbs.dm:575]: Same as add.

File: code/modules/admin/admin_verbs.dm [hunk 4]
```diff
@@ -555,6 +555,27 @@ ADMIN_VERB_ADD(/client/proc/man_up, R_ADMIN, FALSE)
...
+	T.stats.removePerk(perkname.type)
```
Comment (KathyRyals) [inline code/modules/admin/admin_verbs.dm:576]: Same.

File: code/modules/admin/view_variables/topic.dm [hunk 5]
```diff
@@ -229,6 +229,33 @@
...
 		href_list["datumrefresh"] = href_list["mark_object"]
 
+	else if(href_list["perkadd"])
+		if(!check_rights(0))
+			return
+		var/mob/S = locate (href_list["perkadd"])
+		if(!istype(S))
+			to_chat(usr, "This can only be done to instances of type /mob")
+			return
+		var/datum/perk/perkname = input("What perk do you want to add?") in subtypesof(/datum/perk/)
```
Comment (KathyRyals) [inline code/modules/admin/view_variables/topic.dm:239]: Needs to be cancelable.

File: code/modules/admin/view_variables/topic.dm [hunk 6]
```diff
@@ -229,6 +229,33 @@
...
+		S.stats.addPerk(perkname)
```
Comment (KathyRyals) [inline code/modules/admin/view_variables/topic.dm:240]: Same as `T` previously. Need to sanity check it.

File: code/modules/admin/view_variables/topic.dm [hunk 7]
```diff
@@ -229,6 +229,33 @@
...
+		if(!check_rights(0))
+			return
+		var/mob/S = locate (href_list["perkremove"])
+		if(!istype(S))
+			to_chat(usr, "This can only be done to instances of type /mob")
+			return
+		if (S.stats.perks.len ==0)
+			to_chat(usr, "Creature has no perks to remove")
+			return
+		var/datum/perk/perkname = input("What perk do you want to remove?") in S.stats.perks
```
Comment (KathyRyals) [inline code/modules/admin/view_variables/topic.dm:254]: You know by now.

File: code/modules/admin/view_variables/topic.dm [hunk 8]
```diff
@@ -229,6 +229,33 @@
...
+		S.stats.removePerk(perkname.type)
```
Comment (KathyRyals) [inline code/modules/admin/view_variables/topic.dm:255]: Idem.

**Review by KathyRyals — CHANGES_REQUESTED** [review]

**Comment by MSieghardt** [comment]
Changed applied and tested, you can now cancel out of it. Sanity check added to close out before we call addPerk or removePerk.

**Comment by Kazkin** [comment]
Nothing jumps out at me as out of place with the changes. I'll wait for Lessie to give it a check too before merging.

**Comment by KathyRyals** [comment]
He hasn't pushed the required changes yet.

**Comment by MSieghardt** [comment]
Oh, my fault, thought I did.

**Review by KathyRyals — APPROVED** [review]

**Pull request merged**